### PR TITLE
Feature/variable monitoring messages

### PR DESCRIPTION
**Pull request by AssemblyJohn**
## Describe your changes

Handling of monitoring messages with dispatch to the device model.

## Issue ticket number and link

Depends on: https://github.com/EVerest/libocpp/pull/653
Issue: https://github.com/EVerest/libocpp/issues/640

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: lib/ocpp/v201/charge_point.cpp [hunk 1]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+
+            // TODO(ioan): Spec 3.56: see if we need to clear custom monitors (the ones
+            // set by the CSMS) on 'FactoryDefault' too
+            if (msg.monitoringBase == MonitoringBaseEnum::HardWiredOnly) {
+                try {
+                    this->device_model->clear_custom_monitors();
+                } catch (const QueryExecutionException& e) {
+                    EVLOG_warning << "Could not clear custom monitors from DB: " << e.what();
+                    response.status = GenericDeviceModelStatusEnum::Rejected;
+                }
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3246]: The device model storage implementation currently doesn't throw `QueryExecutionException`, this is to be added in a subsequent PR 

We have an abstraction layer for the DeviceModelStorage, so we probably do not want to define Device Model specific exceptions for these cases

File: lib/ocpp/v201/charge_point.cpp [hunk 2]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+
+void ChargePoint::handle_set_variable_monitoring_req(Call<SetVariableMonitoringRequest> call) {
+    SetVariableMonitoringResponse response;
+    const auto& msg = call.msg;
+
+    try {
+        response.setMonitoringResult = this->device_model->set_monitors(msg.setMonitoringData);
+    } catch (const ocpp::common::QueryExecutionException& e) {
+        EVLOG_error << "Set monitors failed:" << e.what();
+    }
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3289]: Same here

File: lib/ocpp/v201/charge_point.cpp [hunk 3]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+
+        if (!data.empty()) {
+            response.status = GenericDeviceModelStatusEnum::Accepted;
+        } else {
+            response.status = GenericDeviceModelStatusEnum::EmptyResultSet;
+        }
+    } catch (const QueryExecutionException& e) {
+        EVLOG_error << "Get variable monitoring failed:" << e.what();
+        response.status = GenericDeviceModelStatusEnum::Rejected;
+    }
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3359]: Same here

File: lib/ocpp/v201/charge_point.cpp [hunk 4]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+
+void ChargePoint::handle_clear_variable_monitoring_req(Call<ClearVariableMonitoringRequest> call) {
+    ClearVariableMonitoringResponse response;
+    const auto& msg = call.msg;
+
+    try {
+        response.clearMonitoringResult = this->device_model->clear_monitors(msg.id);
+    } catch (const ocpp::common::QueryExecutionException& e) {
+        EVLOG_error << "Clear variable monitoring failed:" << e.what();
+    }
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3378]: Same here

File: lib/ocpp/v201/device_model_storage_sqlite.cpp [hunk 5]
```diff
@@ -367,6 +367,23 @@ bool DeviceModelStorageSqlite::clear_variable_monitor(int monitor_id) {
     return (delete_stmt->changes() == 1);
 }
 
+int32_t DeviceModelStorageSqlite::clear_custom_variable_monitors() {
```
Comment (Pietfried) [inline lib/ocpp/v201/device_model_storage_sqlite.cpp:370]: Please add a test for `clear_custom_variable_monitors`

File: lib/ocpp/v201/charge_point.cpp [hunk 6]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+    const auto& msg = call.msg;
+
+    bool valid_monitor = false;
+    switch (msg.monitoringBase) {
+    case MonitoringBaseEnum::All:
+    case MonitoringBaseEnum::FactoryDefault:
+    case MonitoringBaseEnum::HardWiredOnly:
+        valid_monitor = true;
+        break;
+    }
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3223]: Is the `msg.monitoringBase` ever not valid?

File: lib/ocpp/v201/charge_point.cpp [hunk 7]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+
+        if (result != SetVariableStatusEnum::Accepted) {
+            EVLOG_warning << "Could not persist in device model new monitoring base: "
+                          << conversions::monitoring_base_enum_to_string(msg.monitoringBase);
+            response.status = GenericDeviceModelStatusEnum::Rejected;
+        } else {
+            response.status = GenericDeviceModelStatusEnum::Accepted;
+
+            // TODO(ioan): Spec 3.56: see if we need to clear custom monitors (the ones
+            // set by the CSMS) on 'FactoryDefault' too
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3239]: Yes we should, see N03.FR.04 . This was also clarified with OCA

File: lib/ocpp/v201/charge_point.cpp [hunk 8]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+
+            ocpp::Call<NotifyMonitoringReportRequest> call(req, this->message_queue->createMessageId());
+            this->send<NotifyMonitoringReportRequest>(call);
+
+            sequence_num++;
+        }
+    }
+}
+
+void ChargePoint::handle_get_variable_monitoring_req(Call<GetMonitoringReportRequest> call) {
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3341]: function name should be `handle_get_monitoring_report_req`

File: lib/ocpp/v201/charge_point.cpp [hunk 9]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+    GetMonitoringReportResponse response;
+    const auto& msg = call.msg;
+
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3344]: You can implement `N02.FR.07` similar to https://github.com/EVerest/libocpp/blob/main/lib/ocpp/v201/charge_point.cpp#L2295

File: lib/ocpp/v201/charge_point.cpp [hunk 10]
```diff
@@ -3194,6 +3209,178 @@ void ChargePoint::handle_customer_information_req(Call<CustomerInformationReques
...
+    }
+
+    ocpp::CallResult<SetMonitoringLevelResponse> call_result(response, call.uniqueId);
+    this->send<SetMonitoringLevelResponse>(call_result);
+}
+
+void ChargePoint::handle_set_variable_monitoring_req(Call<SetVariableMonitoringRequest> call) {
+    SetVariableMonitoringResponse response;
+    const auto& msg = call.msg;
+
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:3284]: You can implement N04.FR.09 similar to https://github.com/EVerest/libocpp/blob/main/lib/ocpp/v201/charge_point.cpp#L2295

**Review by Pietfried — CHANGES_REQUESTED** [review]

**Comment by AssemblyJohn** [comment]
All comments have been implemented.

As a note, due to heavy modifications, almost all the modifications in the production-preparation branch probably have to be re-done (a rebase would be too difficult with so many divergent implementation parts): https://github.com/EVerest/libocpp/tree/feature/variable-monitoring-messages-notifications

My suggestion would be to freeze other spec-based fixes, and move them to the final production branch.

**Review by Pietfried — CHANGES_REQUESTED** [review]
Just a couple of minor things

File: lib/ocpp/v201/charge_point.cpp [hunk 11]
```diff
@@ -409,7 +409,7 @@ void ChargePoint::on_transaction_finished(const int32_t evse_id, const DateTime&
         if (meter_values.value().empty()) {
             meter_values.reset();
         }
-    } catch (const QueryExecutionException& e) {
+    } catch (const DeviceModelStorageError& e) {
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply lib/ocpp/v201/charge_point.cpp]: I don't understand the request, we're catching a DatabaseException there: https://github.com/EVerest/libocpp/blob/7d456bfec6105eb3dbb7edba3dafae22bf0d2ef1/lib/ocpp/v201/charge_point.cpp#L412

File: lib/ocpp/v201/charge_point.cpp [hunk 12]
```diff
@@ -3257,13 +3460,15 @@ void ChargePoint::handle_send_local_authorization_list_req(Call<SendLocalListReq
                     auto entries = this->database_handler->get_local_authorization_list_number_of_entries();
                     this->device_model->set_read_only_value(local_entries.component, local_entries.variable.value(),
                                                             AttributeEnum::Actual, std::to_string(entries));
-                } catch (const QueryExecutionException& e) {
-                    EVLOG_warning << "Could not get local list count from database";
+                } catch (const DeviceModelStorageError& e) {
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply lib/ocpp/v201/charge_point.cpp:3463]: It seems that in the main there has been added a transaction that can throw a 'DeviceModelStorageError': https://github.com/EVerest/libocpp/blob/9d605a925e494e8e8893559e4fe3ef182a9705ef/lib/ocpp/v201/device_model_storage_sqlite.cpp#L209

File: lib/ocpp/v201/charge_point.cpp [hunk 13]
```diff
@@ -17,7 +17,7 @@ using namespace std::chrono_literals;
 
 const auto DEFAULT_MAX_CUSTOMER_INFORMATION_DATA_LENGTH = 51200;
 
-using QueryExecutionException = ocpp::common::QueryExecutionException;
+using DatabaseException = ocpp::common::DatabaseException;
```
[the remark this answers is not in the document]
Comment (AssemblyJohn) [reply lib/ocpp/v201/charge_point.cpp:20]: There seem to be more places where this is used, like the `message_queue.hpp`. In principle still using the `QueryExecException` should not change any of the current functionality.

**Pull request merged**